src/DominantJuice.sol

```diff
@@ -0,0 +1,399 @@
+ * extends the functionality of a Juicebox project. Once a JB project is created, the constructor stores project data
+ * and the pertinent Juicebox architecture contracts. This contract is written for a single cycle campaign. If projects
+ * want dominant assurance on more cycles, refactoring is required.
+ */
+contract DominantJuice is
+    IJBFundingCycleDataSource3_1_1,
+    IJBPayDelegate3_1_1,
+    IJBRedemptionDelegate3_1_1,
+    AccessControl
+{
+    bytes32 private constant CAMPAIGN_MANAGER_ROLE = keccak256("CAMPAIGN_MANAGER_ROLE");
+
+    /// Juicebox Contracts
+    IJBController3_1 public immutable controller;
+    IJBDirectory public immutable directory; // directory of terminals and controllers for projects.
+    IJBFundAccessConstraintsStore public immutable fundAccessConstraintsStore;
+    IJBSingleTokenPaymentTerminalStore3_1_1 public immutable paymentTerminalStore;
+    IJBSingleTokenPaymentTerminal public immutable paymentTerminal;
+
+    /// Main Campaign Parameters
+    struct Campaign {
+        uint256 projectId;
+        uint256 cycleTarget;
+        uint256 cycleExpiryDate;
+        uint256 minimumPledgeAmount;
+        uint256 totalRefundBonus;
+    }
+
+    struct FundingStatus {
+        uint256 totalPledged;
+        uint256 percentOfGoal;
+        bool isTargetMet;
+        bool hasCycleExpired;
+        bool hasCreatorWithdrawnAllFunds;
+    }
+
+    struct Pledgers {
+        uint256 totalAmountPledged;
+        UD60x18 totalPledgeWeight;
+        mapping(address => UD60x18) pledgerWeight;
+        mapping(address => bool) hasBeenRefunded;
+    }
+
+    /// State Variables
+    Campaign private campaign;
+    Pledgers private pledgers;
+    UD60x18 private immutable rateOfDecay = ud(0.99e18);
+
+    // Events for transparency to pledgers
+    event RefundBonusDeposited(address, uint256 indexed);
+    event PledgeMade(address, uint256);
+    event CycleHasClosed(bool indexed, bool indexed);
+    event CycleRefundBonusWithdrawal(address indexed, uint256 indexed);
+    event CreatorWithdrawal(address, uint256);
+
+    error BonusAlreadyDeposited(uint256 bonusAmount);
+    error RefundBonusNotDeposited();
+    error AmountIsBelowMinimumPledge(uint256 minAmount);
+    error InvalidPaymentEvent(address caller, uint256 projectId, uint256 value);
+    error CycleHasExpired();
+    error IncorrectProjectID();
+    error FunctionHasAlreadyBeenCalled();
+    error MustBePledger();
+    error CycleHasNotEndedYet(uint256 endTimestamp);
+    error NoRefundsForSuccessfulCycle();
+    error AlreadyWithdrawnRefund();
+    error InvalidRedemptionEvent(address caller, uint256 projectId, uint256 value);
+    error InsufficientFunds();
+
+    // Make sure the caller is a terminal of the project, and that the call is being made
+    // on behalf of an interaction with the correct project.
+    modifier terminalCheck() {
+        require(msg.value == 0, "Pledges should be made through JB website.");
```

Comment:
Any reasons not to use custom errors here?

Reply:
I don’t know if there is a magical number to where custom errors get too long, but something like `CycleIsActiveHasFailedOrFunctionIsLocked()` felt too long, or you would need commas or other syntax to make it readable. If you think of a short one that conveys all that info, please send on over!

Reply:
Will add. I wasn’t sure if you would want this modifier changed after the update, and so I’m trying to ride that line between getting things too polished and having them convey the nuts and bolts to see if you want to keep or not. I gotta make every hour count over here, so I’m just trying to streamline / not get ahead of myself.

Reply:
no worries. feel free to keep this as is.